src/threading/cthread.hpp

```diff
@@ -0,0 +1,303 @@
+        return f();
+    }
+
+public:
+    parallel_vector() = default;
+    using iterator = typename std::vector<value_type>::iterator;
+    using const_iterator = typename std::vector<value_type>::const_iterator;
+
+    iterator begin() { return data_.begin(); }
+    iterator end()   { return data_.end(); }
+
+    const_iterator begin() const { return data_.begin(); }
+    const_iterator end()   const { return data_.end(); }
+
+    const_iterator cbegin() const { return data_.cbegin(); }
+    const_iterator cend()   const { return data_.cend(); }
+
+    void push_back (value_type&& val) {
+        critical([&] {
+            data_.push_back(std::move(val));
+        });
+    }
+};
+
+inline std::string description() {
+    return "CThread Pool";
+}
+
+struct timer {
+    using time_point = std::chrono::time_point<std::chrono::system_clock>;
+
+    static inline time_point tic() {
+        return std::chrono::system_clock::now();
+    }
+
+    static inline double toc(time_point t) {
+        return std::chrono::duration<double>{tic() - t}.count();
+    }
+
+    static inline double difference(time_point b, time_point e) {
+        return std::chrono::duration<double>{e-b}.count();
+    }
+};
+
+constexpr bool multithreaded() { return true; }
+
+class task_group {
+private:
+    std::size_t in_flight = 0;
+    impl::task_pool& global_task_pool;
+    // task pool manipulates in_flight
+    friend impl::task_pool;
+  
+public:
+    task_group():
+        global_task_pool{impl::task_pool::get_global_task_pool()}
+    {}
+    
+    task_group(const task_group&) = delete;
+    task_group& operator=(const task_group&) = delete;
+
+    // send function void f() to threads
+    template<typename F>
+    void run(const F& f) {
+        global_task_pool.run(impl::task{f, this});
+    }
+
+    template<typename F>
+    void run(F&& f) {
+        global_task_pool.run(impl::task{std::move(f), this});
+    }
+
+    // run function void f() and then wait on all threads in group
+    template<typename F>
+    void run_and_wait(const F& f) {
+        f();
+        global_task_pool.wait(this);
+    }
+
+    template<typename F>
+    void run_and_wait(F&& f) {
+        f();
+        global_task_pool.wait(this);
+    }
+
+    // wait till all tasks in this group are done
+    void wait() {
+        global_task_pool.wait(this);
+    }
+
+    // Make sure that all tasks are done before clean up
+    ~task_group() {
+        wait();
+    }
+};
+
+///////////////////////////////////////////////////////////////////////
+// algorithms
+///////////////////////////////////////////////////////////////////////
+struct parallel_for {
+    template <typename F>
+    static void apply(int left, int right, F f) {
+        task_group g;
+        for(int i = left; i < right; ++i) {
+          g.run([=] {f(i);});
+        }
+        g.wait();
+    }
+};
+
+} // threading
+} // mc
+} // nest
+
+// paralele stable sort uses threading
+#include "cthread_parallel_stable_sort.h"
+
```

Comment:
Suggestion? It is confusing, but it is dependent on the definitions in cthread.hpp, and the sorts depend on it. So the other way is to split all of this up into cthread_impl, cthread_sort and cthread_parallel_stable_sort, which maybe what should be done... but on the other hand, it adds a different type of complexity.